### ISSUE TITLE
代码是否不完全？

### ISSUE THREAD

**Issue by LiLuoQiu** [Magnety/Multi_modal_Image#2]
我在运行时，发现arch.BigGAN.utils路径是没有文件的，并且show_grid和focal loss文件也没有，focal loss我能够构建，但是这些文件我并不知道其具体内容。此外，我想单纯运行一下该模型网络在我的数据集中进行训练，能否麻烦你们给一份readme介绍文件或者其他说明类的文件？十分感谢

**Comment by Magnety** [comment]
arch.BigGAN.utils 非必要，
show_grid代码如下：
import numpy as np
import SimpleITK as sitk
import matplotlib.pyplot as plt
from scipy.ndimage.interpolation import map_coordinates

import torch
import torch.nn as nn
import torch.nn.functional as F
import io
import PIL

class SpatialTransformer(nn.Module):
    def __init__(self, size, mode='bilinear'):
        """
        Instiatiate the block
            :param size: size of input to the spatial transformer block
            :param mode: method of interpolation for grid_sampler
        """
        super(SpatialTransformer, self).__init__()
        # Create sampling grid
        vectors = [torch.arange(0, s) for s in size]
        grids = torch.meshgrid(vectors)
        grid = torch.stack(grids)  # y, x, z
        grid = torch.unsqueeze(grid, 0)  # add batch
        grid = grid.type(torch.FloatTensor)
        self.register_buffer('grid', grid)

        self.mode = mode

    def forward(self, src, flow):
        """
        Push the src and flow through the spatial transform block
            :param src: the original moving image
            :param flow: the output from the U-Net
        """
        new_locs = self.grid + flow

        shape = flow.shape[2:]

        # Need to normalize grid values to [-1, 1] for resampler
        for i in range(len(shape)):
            new_locs[:, i, ...] = 2 * (new_locs[:, i, ...] / (shape[i] - 1) - 0.5)

        if len(shape) == 2:
            new_locs = new_locs.permute(0, 2, 3, 1)  
            new_locs = new_locs[..., [1, 0]]
        elif len(shape) == 3:
            new_locs = new_locs.permute(0, 2, 3, 4, 1)
            new_locs = new_locs[..., [2, 1, 0]]

        return F.grid_sample(src, new_locs, mode=self.mode,align_corners=True)
def grid_map(grid,out_path):
    phi_arr = grid
    size = phi_arr.shape
    num1, num2 = (size[0] + 10) // 5, (size[1] + 10) // 5  # 改变除数（10），即可改变网格的密度
    x, y = np.meshgrid(np.linspace(-2, 2, num1), np.linspace(-2, 2, num2))
    plt.figure(figsize=((size[0] *2) / 100.0, (size[1] *2) / 100.0))  # 指定图像大小
    plt.plot(x, y, color="black")
    plt.plot(x.transpose(), y.transpose(), color="black")
    plt.axis('off')  
    plt.gca().xaxis.set_major_locator(plt.NullLocator())
    plt.gca().yaxis.set_major_locator(plt.NullLocator())
    plt.subplots_adjust(top=1, bottom=0, left=0, right=1, hspace=0, wspace=0)
    plt.margins(0, 0)
    buffer_ = io.BytesIO()
    plt.savefig(buffer_,format='png')  
    dataPIL = PIL.Image.open(buffer_)
    data = np.asarray(dataPIL)
    img = data[..., 0]
    buffer_.close()
    img = np.squeeze(img)[np.newaxis, np.newaxis, :size[0], :size[1]]
    STN = SpatialTransformer(size[:2])
    phi_arr = phi_arr.permute(2, 0, 1)[np.newaxis, ...]
    warp = STN(torch.from_numpy(img).float(), phi_arr)
    warp_img = sitk.GetImageFromArray(warp[0, 0, ...].numpy().astype(np.uint8))
    sitk.WriteImage(warp_img, out_path)
    plt.close()
def grid_map_1(grid,out_path):
    W,H,D = grid.shape
    x, y = np.meshgrid(np.arange(0, W, 1), np.arange(0, H, 1))
    u = grid[:, :, 0]
    v = grid[:, :, 1]
    plt.quiver(x[::5,::5], y[::5,::5], u[::5,::5], v[::5,::5], scale=1,color='red')
    plt.savefig(out_path)
    plt.close()

**Comment by LiLuoQiu** [comment]
> arch.BigGAN.utils 非必要， show_grid代码如下： import numpy as np import SimpleITK as sitk import matplotlib.pyplot as plt from scipy.ndimage.interpolation import map_coordinates
> 
> import torch import torch.nn as nn import torch.nn.functional as F import io import PIL
> 
> class SpatialTransformer(nn.Module): def **init**(self, size, mode='bilinear'): """ Instiatiate the block :param size: size of input to the spatial transformer block :param mode: method of interpolation for grid_sampler """ super(SpatialTransformer, self).**init**() # Create sampling grid vectors = [torch.arange(0, s) for s in size] grids = torch.meshgrid(vectors) grid = torch.stack(grids) # y, x, z grid = torch.unsqueeze(grid, 0) # add batch grid = grid.type(torch.FloatTensor) self.register_buffer('grid', grid)
> 
> ```
>     self.mode = mode
> 
> def forward(self, src, flow):
>     """
>     Push the src and flow through the spatial transform block
>         :param src: the original moving image
>         :param flow: the output from the U-Net
>     """
>     new_locs = self.grid + flow
> 
>     shape = flow.shape[2:]
> 
>     # Need to normalize grid values to [-1, 1] for resampler
>     for i in range(len(shape)):
>         new_locs[:, i, ...] = 2 * (new_locs[:, i, ...] / (shape[i] - 1) - 0.5)
> 
>     if len(shape) == 2:
>         new_locs = new_locs.permute(0, 2, 3, 1)  
>         new_locs = new_locs[..., [1, 0]]
>     elif len(shape) == 3:
>         new_locs = new_locs.permute(0, 2, 3, 4, 1)
>         new_locs = new_locs[..., [2, 1, 0]]
> 
>     return F.grid_sample(src, new_locs, mode=self.mode,align_corners=True)
> ```
> 
> def grid_map(grid,out_path): phi_arr = grid size = phi_arr.shape num1, num2 = (size[0] + 10) // 5, (size[1] + 10) // 5 # 改变除数（10），即可改变网格的密度 x, y = np.meshgrid(np.linspace(-2, 2, num1), np.linspace(-2, 2, num2)) plt.figure(figsize=((size[0] *2) / 100.0, (size[1] *2) / 100.0)) # 指定图像大小 plt.plot(x, y, color="black") plt.plot(x.transpose(), y.transpose(), color="black") plt.axis('off') plt.gca().xaxis.set_major_locator(plt.NullLocator()) plt.gca().yaxis.set_major_locator(plt.NullLocator()) plt.subplots_adjust(top=1, bottom=0, left=0, right=1, hspace=0, wspace=0) plt.margins(0, 0) buffer_ = io.BytesIO() plt.savefig(buffer_,format='png') dataPIL = PIL.Image.open(buffer_) data = np.asarray(dataPIL) img = data[..., 0] buffer_.close() img = np.squeeze(img)[np.newaxis, np.newaxis, :size[0], :size[1]] STN = SpatialTransformer(size[:2]) phi_arr = phi_arr.permute(2, 0, 1)[np.newaxis, ...] warp = STN(torch.from_numpy(img).float(), phi_arr) warp_img = sitk.GetImageFromArray(warp[0, 0, ...].numpy().astype(np.uint8)) sitk.WriteImage(warp_img, out_path) plt.close() def grid_map_1(grid,out_path): W,H,D = grid.shape x, y = np.meshgrid(np.arange(0, W, 1), np.arange(0, H, 1)) u = grid[:, :, 0] v = grid[:, :, 1] plt.quiver(x[::5,::5], y[::5,::5], u[::5,::5], v[::5,::5], scale=1,color='red') plt.savefig(out_path) plt.close()

你好，非常感谢你的解惑。我对你们的研究十分感兴趣，但还是存在一定的疑惑，请问如果我想直接调用你们“ours”模型，其中的cam参数不使用.BigGAN.构建，还可以使用其他方式输入吗？如果我需要根据你们研究进行下游任务适配，我可以怎么获取cam参数，以及我的数据是否应该做出一定程度的适配改变呢？十分感谢你在忙碌中对我的问题进行解答。
